SSA/Projects/FullyHomomorphicEncryption/PrettySyntax.lean

```diff
@@ -0,0 +1,53 @@
+import SSA.Core.MLIRSyntax.PrettyEDSL
+import SSA.Projects.FullyHomomorphicEncryption.Syntax
+
+import Mathlib.Init.Data.Nat.Lemmas
+
+namespace MLIR.EDSL.Pretty
+open Lean
+
+syntax "poly.add" : MLIR.Pretty.uniform_op
+syntax "return" : MLIR.Pretty.uniform_op
+
+syntax mlir_op_operand " = " "arith.const" "$" noWs "{" term "}" " : " mlir_type : mlir_op
+syntax mlir_op_operand " = " "arith.const" neg_num " : " mlir_type : mlir_op
+macro_rules
+  | `(mlir_op| $v:mlir_op_operand = arith.const $x:neg_num : $t) =>
+      `(mlir_op| $v:mlir_op_operand = "arith.const" () {value = $x:neg_num } : () -> ($t))
+  | `(mlir_op| $v:mlir_op_operand = arith.const ${ $x:term } : $t) => do
+      let ctor := mkIdent ``MLIR.AST.AttrValue.int
+      let x ← `($ctor $x [mlir_type| i64])
+      --                         ^^^^^^^^^^^^^^^^ Is this right? It seems we just ignore the
+      --                                          type annotation
```

Comment:
I don't understand this, I would have assumed we use `$t`?

Reply:
So that doesn't work, because we sometimes have `arith.const`s of type `index`, and something somewhere expects only `i<num>` or `f<num>`.

I took this from the examples currently in `FHE/Rewrites.lean`, e.g.:
```lean
noncomputable def lhs := [fhe_com q, n, hq| {
^bb0(%a : ! R):
  %one_int = "arith.const" () {value = 1}: () -> (i16)
  %two_to_the_n = "arith.const" () {value = $((2**n : Int))}: () -> (index)
  %x2n = "poly.monomial" (%one_int,%two_to_the_n) : (i16, index) -> (! R)
  %oner = "poly.const" () {value = 1}: () -> (! R)
  %p = "poly.add" (%x2n, %oner) : (! R, ! R) -> (! R)
  %v1 = "poly.add" (%a, %p) : (! R, ! R) -> (! R)
  "return" (%v1) : (! R) -> ()
}]
```
Here, the `value` attribute of `arith.const` is *not* given a type attribute, which means it default to `i64` (it does *not* take the type from the type annotation of the op as a whole).
I copied that behaviour, but with some reservations, hence, the comment